Install/Remove/Uninstall from Command line without GUI

tarTG opened the pull request:
Fixes #2907

Currently it works like
**Silent Install**
lutris --install (game slug / url) --bin_path path_to_binary1; path_to_binary2 [--install_path path_to_install_folder]

If install folder is not given, the default one is used.


**Silent Remove**
lutris --remove id/game_slug

**Silent Uninstall**
lutris --uninstall id/game_slug

Missing Runners are installed without asking
Disks are searched in the standard search paths

I don't know how to handle input menus for this case. So the program prints an error and stops.
Any suggestions?

AlexanderRavenheart commented on lutris/installer/silient_install.py:
```diff
@@ -0,0 +1,219 @@
+            self.interpreter = interpreter.ScriptInterpreter(install_script, self)
+        except MissingGameDependency as ex:
+            # call recursive dependencies
+            SilenInstall(
```
Missing t from Silent

AlexanderRavenheart commented on lutris/installer/silient_install.py:
```diff
@@ -0,0 +1,219 @@
+from lutris.util.downloader import Downloader
+
+
+class SilenInstall:
```
Missing t from Silent

AlexanderRavenheart commented on lutris/installer/silient_install.py:
```diff
@@ -0,0 +1,219 @@
+        self._print(self.commandline, text)
+
+    # required by interpreter
+    def clean_widgets(self):
```
Required? This SilentInstall class does not extend or inherit any class, so where is this requirement coming from?

tarTG replied:
The [Script Interpreter](https://github.com/lutris/lutris/blob/155dc4e61ac9e65bec0deb24e8fe27faa200caf4/lutris/installer/interpreter.py#L88) requires a parent class. At the moment this is only the installer window and via this function, the script interpreter make changes on the Window.

Either there had to be a lot of changes in the script interpreter (like raising signals instead of calling the installer window as parent) or the GUI functions have to be available, so the program does not crash.
This is also the reason why there are some changes to GUI accesses like
`continue_button.hide()` 
into functions
`continue_button_hide()`
Otherwise there would be a lot of GUI dependencies in the silent installer.

AlexanderRavenheart commented on lutris/installer/silient_install.py:
```diff
@@ -0,0 +1,219 @@
+        pass
+
+    # required by interpreter
+    def add_spinner(self):
```
Required? This SilentInstall class does not extend or inherit any class, so where is this requirement coming from?

AlexanderRavenheart commented on lutris/installer/silient_install.py:
```diff
@@ -0,0 +1,219 @@
+        pass
+
+    # required by interpreter
+    def set_cancle_butten_sensitive(self):
```
Required? This SilentInstall class does not extend or inherit any class, so where is this requirement coming from?

AlexanderRavenheart commented on lutris/installer/silient_install.py:
```diff
@@ -0,0 +1,219 @@
+        pass
+
+    # required by interpreter
+    def continue_button_hide(self):
```
Required? This SilentInstall class does not extend or inherit any class, so where is this requirement coming from?

AlexanderRavenheart commented on lutris/installer/silient_install.py:
```diff
@@ -0,0 +1,219 @@
+        pass
+
+    def attach_logger(self, command):
+        pass
```
Why does this function even exist?

AlexanderRavenheart commented:
If the actions are silent then there shouldn't be any input menus or other user interaction required as that kinda defeats the purpose of being silent, right?

tarTG commented:
I invoke the install process, like the installwindow. 
But the process and the installwindow are not decoupled and invoke each other.

The GUI functions (addSpinner etc.) are required, because of the direct GUI accesses. 
Without this, i have to refactor a lot of function within the original interpreter, the installerwindow and the commands files.

tarTG commented on lutris/installer/interpreter.py:
```diff
@@ -492,7 +492,7 @@ def get_runner_class(self, runner_name):
             runner = import_runner(runner_name)
         except InvalidRunner:
-            GLib.idle_add(self.parent.cancel_button.set_sensitive, True)
+            self.parent.set_cancle_butten_sensitive()
```
I did not. 
I just moved it into the function of the installer window, so the silent installer has no need for having a cancel button but just an empty function

strycore commented on lutris/installer/commands.py:
```diff
@@ -216,7 +216,7 @@ def insert_disc(self, data):
         )
         if self.runner == "wine":
-            GLib.idle_add(self.parent.eject_button.show)
+            self.parent.eject_button_show()
```
why did you remove that,

tarTG replied:
I moved it into the installer window.
https://github.com/lutris/lutris/pull/3029/files#diff-09064015f83e9cec3fca291ce9ae5954R538-R539

Otherwise the silent installer needs an eject_button with all the GUI dependencies.
This way I only need to provide an empty function.

strycore commented:
Input menu values have to be passed from the command line. The unattended installer must check that all values have been provided before launching the install and quit if one is not provided/invalid.

Also, I don't think the remove command line flag should be added.

tarTG commented:
Deleted remove command line option.
Added -input_menu option:option... for menu choices.
Check all arguments before installing.
Multiple arguments now divided by **:** because **;** terminates command line statements.
Fixed errors and typos.

The only thing which bothers me, is that installers who don't need any files are installed with 
`./lutris --install installer --bin_path (random_argument) `
The argument never gets evaluated but it is really ugly.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/lutris/lutris?pullRequest=3029) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/lutris/lutris?pullRequest=3029) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/lutris/lutris?pullRequest=3029) it.</sub>

kirtr commented:
I'd love to see this feature added to Lutris. Thank you for the patches, @tarTG! It would greatly help me to automate some of the game installation for the gaming computers I maintain for our kids and their friends. Any chance of signing the Contributor License Agreement to finish this off?

yoyonel commented on lutris/gui/application.py:
```diff
@@ -395,6 +433,42 @@ def do_command_line(self, command_line):  # noqa: C901  # pylint: disable=argume
+                game_slug=game_slug,
+                installer_file=installer_file,
+                revision=revision,
+                binary_path=options.lookup_value("bin_path").get_string().split(":"),
```
you're check (above) for `install_path` and `input_menu` but for `bin_path` no if statement ... is it safe ?

yoyonel commented on lutris/gui/application.py:
```diff
@@ -395,6 +433,42 @@ def do_command_line(self, command_line):  # noqa: C901  # pylint: disable=argume
+            return 0
+
+        if action == "uninstall":
+            if not db_game or not db_game["id"]:
```
field/key `id` exist necessarly ?

yoyonel commented on lutris/installer/unattended_install.py:
```diff
@@ -0,0 +1,253 @@
+
+        # Check if given files match the number of needed files in the skript. Also test if the files available
+        if "files" in self.script["script"]:
+            req_files = list(filter(lambda file: "N/A" in file[next(iter(file))], self.script["script"]["files"]))
```
list-comprehension is (a bit) more readable:
```python
req_files = [file for self.script["script"]["files"] if "N/A" in file[next(iter(file))]]
```
